File: src/web/archive.ts
// The local session archive — `.intent/raw-sessions/` holds every raw CC log
// that digestion has read (PRD v0.3.2: evidence outlives Claude Code's ~30-day
// purge; every digest stays permanently re-derivable). This module lists that
// directory and joins it against the sessions table by source_hash, so the
// dashboard can show which preserved evidence is digested and which is not.
//
// Pure join logic is separated from fs so it is unit-testable; the fs read is
// fail-safe (a missing or unreadable directory is an empty archive, never an
// error — the archive is an optional local artifact).

import { readdirSync, statSync } from "node:fs";
import { join, basename } from "node:path";

/** A raw session log preserved in .intent/raw-sessions/. */
export interface RawArchiveFile {
  /** File name, e.g. "b9ab1a0c-….jsonl". */
  file: string;
  /** The digest identity: file basename without .jsonl (matches sessions.source_hash). */
  hash: string;
  sizeBytes: number;
  /** ISO timestamp of last modification (re-copied as the log grows). */
  lastModified: string;
}

/** The subset of a sessions row the join needs. */
export interface DigestedSessionRef {
  id: string;
  sourceHash: string | null;
  startedAt: string | null;
}

/** One archive entry as served to the UI. */
export interface ArchiveEntry {
  hash: string;
  file: string;
  sizeBytes: number;
  lastModified: string;
  /** True when a digested session carries this source_hash. */
  digested: boolean;
  /** The digested session's id, for click-through to the record. */
  sessionId: string | null;
  startedAt: string | null;
}

/**
 * Join raw archive files against digested sessions by source_hash.
 * Newest-first by lastModified — the archive reads like the river.
 */
export function joinArchive(
  files: RawArchiveFile[],
  sessions: DigestedSessionRef[],
): ArchiveEntry[] {
  const byHash = new Map<string, DigestedSessionRef>();
  for (const s of sessions) {
    if (s.sourceHash) byHash.set(s.sourceHash, s);
  }
  return files
    .map((f) => {
      const match = byHash.get(f.hash);
      return {
        hash: f.hash,
        file: f.file,
        sizeBytes: f.sizeBytes,
        lastModified: f.lastModified,
        digested: match !== undefined,
        sessionId: match?.id ?? null,
        startedAt: match?.startedAt ?? null,
      };
    })
    .sort((a, b) => b.lastModified.localeCompare(a.lastModified));
}

/**
 * Read the raw-session archive directory. Fail-safe: any fs problem
 * (missing dir, permissions, a file vanishing mid-scan) yields what could
 * be read — worst case an empty list, never a throw.
 */
export function readRawSessionArchive(dir: string): RawArchiveFile[] {
  let names: string[];
  try {
    names = readdirSync(dir);
  } catch {
    return [];
  }
  const files: RawArchiveFile[] = [];
  for (const name of names) {
    if (!name.endsWith(".jsonl")) continue;
    try {
      const st = statSync(join(dir, name));
      if (!st.isFile()) continue;
      files.push({
        file: name,
        hash: basename(name, ".jsonl"),
        sizeBytes: st.size,
        lastModified: st.mtime.toISOString(),
      });
    } catch {
      // a file disappearing between readdir and stat is not an error
    }
  }
  return files;
}

File: backend/quire/journal/archive.py
"""Archive helpers — ported from journal/src/web/archive.ts and
journal/src/pipeline/orchestrator.ts::archiveRawSession.

Reads the raw-session archive from backend/.intent/raw-sessions/ and
joins it against digested sessions by source_hash. Also owns the write
side: archive_raw_session copies a digested .jsonl into the archive.
"""
from __future__ import annotations

from pathlib import Path
from typing import Callable


def get_archive_dir() -> Path:
    """Return the canonical archive directory (backend/.intent/raw-sessions/)."""
    return Path(__file__).parent.parent.parent / ".intent" / "raw-sessions"


def archive_raw_session(
    log_path: Path | str,
    warn: Callable[[str], None] | None = None,
    archive_dir: Path | None = None,
) -> Path | None:
    """Copy a raw session .jsonl into the archive directory.

    Port of orchestrator.ts::archiveRawSession. Claude Code purges its logs
    on a ~30-day clock; the digest must never be the only survivor. Semantics:
    - skip when the destination exists and is >= the source size
    - re-copy when the source has grown (a resumed session)
    - failure-safe: NEVER raises — archive failure must never fail the digest.

    Returns the destination path when a copy happened, None otherwise
    (skipped or failed). `warn` receives the error message on failure.
    """
    try:
        import shutil

        src = Path(log_path)
        dir_ = archive_dir if archive_dir is not None else get_archive_dir()
        dir_.mkdir(parents=True, exist_ok=True)
        dest = dir_ / src.name  # .jsonl filename == CC session id
        src_size = src.stat().st_size
        if dest.exists() and dest.stat().st_size >= src_size:
            return None
        shutil.copy2(str(src), str(dest))
        return dest
    except Exception as err:  # noqa: BLE001 — failure-safe by contract
        if warn is not None:
            try:
                warn(str(err))
            except Exception:
                pass
        return None


def read_raw_session_archive(dir_: Path | str | None = None) -> list[dict]:
    """Read the raw-session archive directory.

    Fail-safe: any filesystem problem returns what could be read — worst case
    an empty list, never a raise.
    """
    d = Path(dir_) if dir_ else get_archive_dir()
    try:
        names = list(d.iterdir())
    except Exception:
        return []
    files = []
    for p in names:
        if not p.name.endswith(".jsonl"):
            continue
        try:
            st = p.stat()
            if not p.is_file():
                continue
            import datetime
            files.append({
                "file": p.name,
                "hash": p.stem,  # basename without .jsonl
                "sizeBytes": st.st_size,
                "lastModified": datetime.datetime.fromtimestamp(
                    st.st_mtime, tz=datetime.timezone.utc
                ).isoformat(),
            })
        except Exception:
            pass
    return files


def join_archive(files: list[dict], sessions: list[dict]) -> list[dict]:
    """Join raw archive files against digested sessions by source_hash.

    Returns entries newest-first by lastModified.
    """
    by_hash: dict[str, dict] = {}
    for s in sessions:
        h = s.get("sourceHash")
        if h:
            by_hash[h] = s

    entries = []
    for f in files:
        match = by_hash.get(f["hash"])
        entries.append({
            "hash": f["hash"],
            "file": f["file"],
            "sizeBytes": f["sizeBytes"],
            "lastModified": f["lastModified"],
            "digested": match is not None,
            "sessionId": match["id"] if match else None,
            "startedAt": match.get("startedAt") if match else None,
        })

    entries.sort(key=lambda e: e["lastModified"], reverse=True)
    return entries
